opt_adapt/opt.py

```diff
@@ -201,13 +201,15 @@ def _gradient_descent(it, forward_run, m, params, u, u_, dJ_):
     P = dJ.copy(deepcopy=True)
     P *= -1
 
-    # Choose step length
-    if u_ is not None and dJ_ is not None:
+    # Choose step length by the Barzilai-Borwein formal
+    if u_ is None or dJ_ is None:
+        lr = params.lr
+    else:
         dJ_ = params.transfer_fn(dJ_, dJ.function_space())
         u_ = params.transfer_fn(u_, u.function_space())
         dJ_diff = fd.assemble(ufl.inner(dJ_ - dJ, dJ_ - dJ) * ufl.dx)
-        params.lr = abs(fd.assemble(ufl.inner(u_ - u, dJ_ - dJ) * ufl.dx) / dJ_diff)
-    lr = line_search(forward_run, m, u, P, J, dJ, params)
+        lr = abs(fd.assemble(ufl.inner(u_ - u, dJ_ - dJ) * ufl.dx) / dJ_diff)
+    lr = max(lr, params.lr_min)
```

Comment:
But I thought the idea was to apply line search to check that we do indeed have a descent direction?

Reply:
Sorry, doctor, I'm not quite sure what you mean. I think

`lr = abs(fd.assemble(ufl.inner(u_ - u, dJ_ - dJ) * ufl.dx) / dJ_diff)` and 
`lr = line_search(forward_run, m, u, P, J, dJ, params)`

are two different ways to find step length. For the code
`lr = max(lr, params.lr_min)`
I just wanna the final step length greater than the machine’s precision.

Reply:
Okay. I just meant surely it doesn't hurt to apply the line search in addition to the Barzilai-Borwein formula.